File: 30_Cube/ruby/cube.rb
$landmines = Array.new

$currentLocation = "111"

$standings = 500 # starting amount

def getYesOrNoResponseTo prompt
    print prompt
    # strip leading and trailing whitespace from entry
    yesno = gets.strip.upcase[0]
    yesno == "Y"
end

def greeting
    puts "CUBE".center(80)
    puts "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY".center(80)
    puts "\n\n\n"
    response = getYesOrNoResponseTo "Do you want to see the INSTRUCTIONS?"
    if response
        puts "This is a game in which you will be playing against the"
        puts "random decision of the computer. The field of play is a"
        puts "cube of size 3. Any of your 27 locations can be designated"
        puts "by inputting three numbers such as 231."
        puts "At the start you are automatically at location 1,1,1.\n"
        puts "The object of the game is to get to location 3,3,3.\n"

        puts "\nONE MINOR DETAIL:"
        puts "The computer will pick five random locations at which it will"
        puts "plant land mines. if you hit one of these locations you lose.\n"

        puts "\nONE OTHER DETAIL:"
        puts "You may move only one space in one direction each move."
        puts "For example: from 1,1,2 you may move to 2,1,2 or 1,1,3."
        puts "You may not change more than one number on the same move."
        puts "If you make an illegal move, you lose and the computer takes"
        puts "the money you may have bet on that round."
        puts ""
        puts "When stating the amount of a wager, enter only the number"
        puts "of dollars (example: 250)  You are automatically started with"
        puts "500 dollars in your account."
        puts
        puts "Good luck!"
    end
end

def landMindStringFrom x, y, z
    landMine = x.to_s + y.to_s + z.to_s
    return landMine
end

def assignLandMines
    $landmines.clear

    # put five unique entries into the landmines array
    while $landmines.size < 5 do
        a = rand(3)+1
        b = rand(3)+1
        c = rand(3)+1
        landmine = landMindStringFrom(a, b, c)
        if !$landmines.include?(landmine) && landmine != "333"
# puts landmine     # debugging
            $landmines.push landmine
        end
    end
    $currentLocation = "111"
end

def initializePot
    $standings = 500 # starting amount
end

def startGame
    assignLandMines
    displayStandings
    response = getYesOrNoResponseTo "WANT TO MAKE A WAGER? "
    if response
        print "HOW MUCH? "
        while true do
        wager = gets.strip.tr('^0-9', '').to_i
            if $standings < wager
                puts "TRIED TO FOOL ME; BET AGAIN ";
            else
                break
            end
        end
    else
        wager = 0
    end

    # start at location 1,1,1
    $currentLocation = "111"
    return wager
end

def goodbye
    puts "TOUGH LUCK!"
    puts ""
    puts "GOODBYE."
    exit
end

def bust
    puts "YOU BUST."
    goodbye
end

def tryAgain
    again = getYesOrNoResponseTo "WANT TO TRY AGAIN? "
    if not again
        exit
    end
end

def isLegalMove? newLocation
    # test for illegal moves
    # can only change one variable per move
    # newLocation is the proposed new position
    # can only move one space from the current position

    moveX = newLocation[0].to_i
    moveY = newLocation[1].to_i
    moveZ = newLocation[2].to_i

    # currentX, currentY, currentZ contains the current position
    currentX = $currentLocation[0].to_i
    currentY = $currentLocation[1].to_i
    currentZ = $currentLocation[2].to_i

    isLegalMove = true
    errorString = ""
   # ensure we're not moving off the cube
    if not (moveX.between?(1,3) && moveY.between?(1,3) && moveZ.between?(1,3))
        errorString = "You moved off the cube!"
        return errorString
    end

    # test for only one move from current position
    if not moveX.between?(currentX-1,currentX+1)
        isLegalMove = false
    end
    if not moveY.between?(currentY-1,currentY+1)
        isLegalMove = false
    end
    if not moveZ.between?(currentZ-1,currentZ+1)
        isLegalMove = false
    end
        if not isLegalMove
            errorString = "You've gone too far"
        end

    # only allow change to one variable at a time
    if isLegalMove
        if moveX != currentX
            if moveY != currentY or moveZ != currentZ
                isLegalMove = false
            end
        end
        if moveY != currentY
            if moveX != currentX or moveZ != currentZ
                isLegalMove = false
            end
        end
        if moveZ != currentZ
            if moveY != currentY or moveX != currentX
                isLegalMove = false
            end
        end
        if not isLegalMove
            errorString = "You made too many changes"
        end
   end

    return errorString
end

def displayStandings
    print "You now have " + $standings.to_s
    if $standings > 1
        puts " dollars"
    else
        puts " dollar"
    end
end

def didWin? location
    location == "333"
end

def youWin amount
    $standings += amount
    puts "*** You win " + amount.to_s + " dollars! ***\n\n"
    displayStandings
    tryAgain
    assignLandMines
    puts "*** new cube ***"
    puts "different landmine locations"
    puts "starting over at location 111"
end

def youLose amount
    # subtract the bet amount from the standings
    if amount > 0
        puts "You lose " + amount.to_s + " dollars!\n\n"
        $standings -= amount
    end
    if $standings <= 0
        # no money left, so end the game
        bust
    else
        displayStandings
    end
    tryAgain
    $currentLocation = "111"
    puts "starting over at location 111"
end

def landMine betAmount
    puts "******BANG******"
    puts "You hit a land mine at " + $currentLocation + "!"
    youLose betAmount
end

def gameLoop betAmount
    while true do
        puts ""
        print "IT'S YOUR MOVE:  "
        # allow only integers: strip anything else from input
        moveToLocation = gets.strip.tr('^0-9', '')

        # test for illegal moves
        # can only change one variable per move
        # moveToLocation is the proposed new position

        error = isLegalMove?(moveToLocation)
        if error == ""
            # assign the new position
            $currentLocation = moveToLocation

            # test for win
            if didWin?(moveToLocation)
                youWin betAmount
            end

            # haven't won yet, test the land mines
            if $landmines.include? moveToLocation
                landMine betAmount
            end

        else
            puts "Illegal move: " + error
            youLose betAmount
        end

    end
end


greeting
initializePot
gameLoop startGame

File: 00_Alternate_Languages/30_Cube/C/cube.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

//check if windows or linux for the clear screen
#ifdef _WIN32
#define CLEAR "cls"
#else
#define CLEAR "clear"
#endif

typedef struct{
    int x;
    int y;
    int z;
}coords;

void instuctions(){
    printf("\nThis is a game in which you will be playing against the\n");
    printf("random decisions of the computer. The field of play is a\n");
    printf("cube of side 3. Any of the 27 locations can be designated\n");
    printf("by inputing three numbers such as 2,3,1. At the start,\n");
    printf("you are automatically at location 1,1,1. The object of\n");
    printf("the game is to get to location 3,3,3. One minor detail:\n");
    printf("the computer will pick, at random, 5 locations at which\n");
    printf("it will plant land mines. If you hit one of these locations\n");
    printf("you lose. One other detail: You may move only one space\n");
    printf("in one direction each move. For example: From 1,1,2 you\n");
    printf("may move to 2,1,2 or 1,1,3. You may not change\n");
    printf("two of the numbers on the same move. If you make an illegal\n");
    printf("move, you lose and the computer takes the money you may\n");
    printf("have bet on that round.\n\n");
    printf("When stating the amount of a wager, printf only the number\n");
    printf("of dollars (example: 250) you are automatically started with\n");
    printf("500 dollars in your account.\n\n");
    printf("Good luck!\n");
}

void game(int money){
    coords player,playerold,mines[5];
    int wager,account = money;
    char choice;
    if(money == 0){
        printf("You have no money left. See ya next time.\n");
        exit(0);
    }
    player.x = 1;
    player.y = 1;
    player.z = 1;
    printf("You have $%d in your account.\n",account);
    printf("How much do you want to wager? ");
    scanf("%d",&wager);
    srand(time(NULL));
    for(int i=0;i<5;i++){
        mines[i].x = rand()%3+1;
        mines[i].y = rand()%3+1;
        mines[i].z = rand()%3+1;
        if(mines[i].x == 3 && mines[i].y == 3 && mines[i].z == 3){
            i--;
        }
    }
    while(player.x != 3 || player.y != 3 || player.z != 3){
        printf("You are at location %d.%d.%d\n",player.x,player.y,player.z);
        if(player.x == 1 && player.y == 1 && player.z == 1)
        printf("Enter new location(use commas like 1,1,2 or else the program will break...): ");
        else printf("Enter new location: ");
        playerold.x = player.x;
        playerold.y = player.y;
        playerold.z = player.z;
        scanf("%d,%d,%d",&player.x,&player.y,&player.z);
        if(((player.x + player.y + player.z) > (playerold.x + playerold.y + playerold.z + 1)) || ((player.x + player.y + player.z) < (playerold.x + playerold.y + playerold.z -1))){
            system(CLEAR);
            printf("Illegal move!\n");
            printf("You lose $%d.\n",wager);
            game(account -= wager);
            break;
        }
        if(player.x < 1 || player.x > 3 || player.y < 1 || player.y > 3 || player.z < 1 || player.z > 3){
            system(CLEAR);
            printf("Illegal move. You lose!\n");
            game(account -= wager);
            break;
        }
        for(int i=0;i<5;i++){
            if(player.x == mines[i].x && player.y == mines[i].y && player.z == mines[i].z){
                system(CLEAR);
                printf("You hit a mine!\n");
                printf("You lose $%d.\n",wager);
                game(account -= wager);
                exit(0);
            }
        }
        if(account == 0){
            system(CLEAR);
            printf("You have no money left!\n");
            printf("Game over!\n");
            exit(0);
        }
    }
    if(player.x == 3 && player.y == 3 && player.z == 3){
        system(CLEAR);
        printf("You made it to the end. You win!\n");
        game(account += wager);
        exit(0);
    }
}

void init(){
    int account = 500;
    char choice;

    printf("Welcome to the game of Cube!\n");
    printf("wanna see the instructions? (y/n): ");
    scanf("%c",&choice);
    if(choice == 'y'){
        system(CLEAR);
        instuctions();
    }
    else if (choice == 'n'){
        system(CLEAR);
        printf("Ok, let's play!\n");
    }
    else{
        system(CLEAR);
        printf("Invalid choice. Try again...\n");
        init();
        exit(0);
    }
    game(account);
    exit(0);
}

void main(){
    init();
}